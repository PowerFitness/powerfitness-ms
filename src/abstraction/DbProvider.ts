import { DbConfig } from '../config'
import { createConnection, Connection, MysqlError } from 'mysql';

export class DbProvider {
	dbConfig: DbConfig;
	connection: Connection | null = null;
	constructor(dbConfig: DbConfig) {
		this.dbConfig = dbConfig;
	}
	establishConnection(): void {
		if(!this.connection) {
			this.connection = createConnection(this.dbConfig)
		}
		return;
	}
	execute<T>(query: string): Promise<T> {
		this.establishConnection();
		return new Promise((resolve, reject) => {
			(this.connection as Connection).query(query, (error: MysqlError, results: T) => {
				if (error) {
					reject(error)
				} else {
					resolve(results)
				}
				(this.connection as Connection).end()
			})
		}) as unknown as Promise<T>;
	}

	beginTransaction(): Promise<void> {
		this.establishConnection();
		return new Promise((resolve, reject) =>
			(this.connection as Connection).beginTransaction((err: MysqlError) => err ? reject(err) : resolve()));
	}

	commit(): Promise<void> {
		this.establishConnection();
		return new Promise((resolve, reject) => (this.connection as Connection).commit((err: MysqlError) => {
			if (err) {
				this.rollback()
					.then(() => reject(err))
					.catch((rollbackError: MysqlError) =>
						reject(new Error(`Rollback Failed: ${err.message} / ${rollbackError.message}`)))
					.finally(() =>
						(this.connection as Connection).end())
			} else {
				resolve();
				(this.connection as Connection).end()
			}
		}));
	}

	rollback(): Promise<void> {
		return new Promise((resolve, reject) =>
			(this.connection as Connection).rollback((err: MysqlError) => err ? reject(err) : resolve()));
	}
}

export default DbProvider;
