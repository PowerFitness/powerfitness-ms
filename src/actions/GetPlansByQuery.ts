import { Action } from '../abstraction/Action';
import { buildWhereAndConditions } from '../util/buildConditions';
import { ParsedQs } from 'qs';
import DbProvider from '../abstraction/DbProvider';

export interface Plan {
	id: number;
	userUniqueId: string;
	motivationStatement: string;
	createdDate: string;
	lastUpdatedDate: string;
}

interface PlansWhereMap { id?: number; userUniqueId?: string; }

export class GetPlansByQuery extends Action<Array<Plan>> {
	queryParams: ParsedQs;
	constructor(dbProvider: DbProvider, queryParams: ParsedQs) {
		super(dbProvider);
		this.queryParams = queryParams;
	}
	buildQuery(): string {
		const { id, userUniqueId } = this.queryParams;
		if (Object.keys(this.queryParams).length === 0) {
			throw new Error('Missing criteria')
		}
		const whereMap: PlansWhereMap = {
			id,
			userUniqueId
		} as PlansWhereMap;

		return `select * from plan where ${buildWhereAndConditions(whereMap)}`;
	}
}

export default GetPlansByQuery;
