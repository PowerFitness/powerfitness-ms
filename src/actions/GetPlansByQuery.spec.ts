import DbProvider from '../abstraction/DbProvider';
import { mock } from 'jest-mock-extended';
import GetPlansByQuery from './GetPlansByQuery';

describe('GetPlansByQuery', () => {
	test('buildQuery', () => {
		const dbProvider: DbProvider = mock<DbProvider>();
		const action: GetPlansByQuery = new GetPlansByQuery(dbProvider, { id: '1',
			userUniqueId: 'uid' });
		expect(action.buildQuery()).toMatchSnapshot();
	})
	test('buildQuery throws an error if no criteria', () => {
		const dbProvider: DbProvider = mock<DbProvider>();
		const action: GetPlansByQuery = new GetPlansByQuery(dbProvider, { });
		expect(action.buildQuery).toThrowError()
	})
})
