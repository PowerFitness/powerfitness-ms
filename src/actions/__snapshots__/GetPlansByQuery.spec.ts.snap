// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetPlansByQuery buildQuery 1`] = `"select * from plan where id = '1' AND userUniqueId = 'uid'"`;
